### PR TITLE
Take automatic game screenshots in Quick Customization

### PR DESCRIPTION
**Pull request by ClementPasteau**
todo: 
- [x] solution for browser preview close
- [ ] solution for mobile app

### CODE REVIEW

File: newIDE/app/src/ExportAndShare/BrowserExporters/BrowserS3PreviewLauncher/index.js [hunk 1]
```diff
@@ -219,10 +230,58 @@ export default class BrowserS3PreviewLauncher extends React.Component<
       );
 
       // If the preview windows are new, register them so that they can be accessed
-      // by the debugger.
+      // by the debugger and for the captures to be detected when they close.
       if (!existingPreviewWindow) {
         previewWindows.forEach((previewWindow: WindowProxy) => {
-          registerNewPreviewWindow(previewWindow);
+          const debuggerId = registerNewPreviewWindow(previewWindow);
+          browserPreviewDebuggerServer.registerCallbacks({
```
Comment (ClementPasteau) [inline newIDE/app/src/ExportAndShare/BrowserExporters/BrowserS3PreviewLauncher/index.js:237]: here I'm using the debugger polling in order to know when the window closes and start the preview screenshot upload logic. I don't know if we're ok with that or if I should create a separate polling logic, or rename the debugger one to a more generic "preview"

Comment (ClementPasteau) [reply newIDE/app/src/ExportAndShare/BrowserExporters/BrowserS3PreviewLauncher/index.js:237]: Also, I find very confusing that callbacks are registered as an array of everything, instead of just adding them one by one

File: newIDE/app/src/MainFrame/index.js [hunk 2]
```diff
@@ -278,13 +281,21 @@ const initialPreviewState: PreviewState = {
   overridenPreviewExternalLayoutName: null,
 };
 
+// Simpler version of the CaptureOptions, as only the timing is needed to start configuring the preview capture.
+type LaunchCaptureOptions = {|
+  screenshots: Array<{|
+    timing: number,
```
Comment (ClementPasteau) [inline newIDE/app/src/MainFrame/index.js:287]: at the moment, the only config for a screenshot is the timing after the preview starts.
It leads to OK results, but maybe we want something more fine-tuned, like a screenshot of each scene ? but then it becomes hard to categorize them well in the backend, as it's just a list of screenshots.

File: newIDE/app/src/MainFrame/index.js [hunk 3]
```diff
@@ -541,6 +552,19 @@ const MainFrame = (props: Props) => {
     onOpenNewProjectSetupDialog: () => setNewProjectSetupDialogOpen(true),
   });
 
+  const {
+    games,
+    gamesFetchingError,
+    fetchGames,
+    onGameUpdated,
+  } = useGamesList();
```
Comment (ClementPasteau) [inline newIDE/app/src/MainFrame/index.js:560]: I've moved this into the mainframe to pass it down, I wanted at first to do the captures logic automatically inside this, but ended up struggling to prevent an effect trigger too many times.
I was that close to create a GamesContext because it's starting to be a mess to pass down games etc everywhere

File: newIDE/electron-app/app/PreviewWindow.js [hunk 4]
```diff
@@ -71,10 +73,11 @@ const openPreviewWindow = ({
 
     previewWindows.push(previewWindow);
 
-    previewWindow.on('closed', event => {
+    previewWindow.on('closed', closeEvent => {
       previewWindows = previewWindows.filter(
         otherPreviewBrowserWindow => otherPreviewBrowserWindow !== previewWindow
       );
+      openEvent.sender.send('preview-window-closed');
```
Comment (ClementPasteau) [inline newIDE/electron-app/app/PreviewWindow.js:80]: locally, I use the event that opened the preview in order to send back a message to the sender. Seems to work well

File: GDJS/GDJS/IDE/ExporterHelper.cpp [hunk 5]
```diff
@@ -768,6 +785,7 @@ void ExporterHelper::AddLibsInclude(bool pixiRenderers,
   InsertUnique(includesFiles, "inputmanager.js");
   InsertUnique(includesFiles, "jsonmanager.js");
   InsertUnique(includesFiles, "Model3DManager.js");
+  InsertUnique(includesFiles, "capturemanager.js");
```
Comment (4ian) [inline GDJS/GDJS/IDE/ExporterHelper.cpp:788]: Would be good to add this only if there is a captureOptions passed. This will avoid yet another file and yet another request when loading (web) games

File: Core/GDCore/IDE/CaptureOptions.h [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+  const long GetTiming() const { return timing; }
+
+  void SetSignedUrl(const gd::String& signedUrl_) { signedUrl = signedUrl_; }
+  const gd::String& GetSignedUrl() const { return signedUrl; }
+
+  void SetPublicUrl(const gd::String& publicUrl_) { publicUrl = publicUrl_; }
+  const gd::String& GetPublicUrl() const { return publicUrl; }
+
+ private:
+  int timing;
```
Comment (4ian) [inline Core/GDCore/IDE/CaptureOptions.h:24]:
```suggestion
  int timing = 0;
```

 Otherwise this has a risk to be initialized to anything that lives in memory here (clang-tidy would probably catch this)

File: GDJS/GDJS/IDE/ExporterHelper.h [hunk 7]
```diff
@@ -48,7 +51,8 @@ struct PreviewExportOptions {
         playerId(""),
         playerUsername(""),
         playerToken(""),
-        allowAuthenticationUsingIframeForPreview(false){};
+        allowAuthenticationUsingIframeForPreview(false),
+        captureOptions({}) {};
```
Comment (4ian) [inline GDJS/GDJS/IDE/ExporterHelper.h:55]: As CaptureOptions is a class, the default constructor will be called, no need to add this line.

File: GDJS/GDJS/IDE/ExporterHelper.h [hunk 8]
```diff
@@ -243,19 +246,34 @@ struct PreviewExportOptions {
...
+  PreviewExportOptions &SetSourceGameId(const gd::String &sourceGameId_) {
     sourceGameId = sourceGameId_;
     return *this;
   }
 
+  /**
+   * \brief Set the capture options to be used for taking screenshots or videos
+   * of the preview.
+   */
+  PreviewExportOptions &AddScreenshotCapture(const int &timing,
```
Comment (4ian) [inline GDJS/GDJS/IDE/ExporterHelper.h:266]: Passing a reference to a int is unusual, it's only when you want to modify it. Passing a const reference means that you can't modify it, but an int in memory is basically the same size as a reference (= pointer).
tl;dr: int can always be passed by value.

```suggestion
  PreviewExportOptions &AddScreenshotCapture(int timing,
```

File: GDJS/Runtime/capturemanager.ts [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+
+    /**
+     * To be called when the scene has started rendering.
+     */
+    setupCaptureOptions(isPreview: boolean): void {
+      if (!isPreview || !this._captureOptions.screenshots) {
+        return;
+      }
+
+      for (const sscreenshotCaptureOption of this._captureOptions.screenshots) {
```
Comment (4ian) [inline GDJS/Runtime/capturemanager.ts:38]:
```suggestion
      for (const screenshotCaptureOption of this._captureOptions.screenshots) {
```

File: GDJS/Runtime/capturemanager.ts [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+        setTimeout(async () => {
+          console.info(
+            `Taking a screenshot and uploading it to ${sscreenshotCaptureOption.publicUrl}`
```
Comment (4ian) [inline GDJS/Runtime/capturemanager.ts:41]:
```suggestion
            `Taking a screenshot and uploading it to ${sscreenshotCaptureOption.publicUrl}.`
```

File: Core/GDCore/IDE/CaptureOptions.h [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+#include "GDCore/String.h"
+
+namespace gd {
+
+class GD_CORE_API Screenshot {
+ public:
+  Screenshot();
+  virtual ~Screenshot() {};
+
+  void SetTiming(int timing_) { timing = timing_; }
```
Comment (4ian) [inline Core/GDCore/IDE/CaptureOptions.h:14]: I would renamed the filed to be "DelayTimeInSeconds" or "DelayTimeInMs"

so that the unit is clear and "Timing" is a bit vague. Delay Time is I think a tad clearer :)

File: GDJS/Runtime/capturemanager.ts [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+          );
+          await this.takeScreenshot(sscreenshotCaptureOption.signedUrl);
+        }, sscreenshotCaptureOption.timing);
+      }
+    }
+
+    /**
+     * Helper function to convert a base64 string to a Blob, which can be uploaded to a server.
+     */
+    base64ToBlob(base64) {
```
Comment (4ian) [inline GDJS/Runtime/capturemanager.ts:51]: uber nitpicking: if it's a helper, you might as well declare it outside of the class so we don't start seeing games using gdjs.CaptureManager.base64ToBlob and complaining when we remove/rename this later. The less API surface, the better.

File: GDJS/Runtime/capturemanager.ts [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+        arrayBuffer[i] = byteString.charCodeAt(i);
+      }
+
+      return new Blob([arrayBuffer], { type: mimeString });
+    }
+
+    /**
+     * Take a screenshot and upload it to the server.
+     */
+    async takeScreenshot(signedUrl: string) {
```
Comment (4ian) [inline GDJS/Runtime/capturemanager.ts:66]:
```suggestion
    async takeAndUploadScreenshot(signedUrl: string) {
```

File: GDJS/Runtime/runtimegame.ts [hunk 14]
```diff
@@ -185,6 +188,11 @@ namespace gdjs {
     _disableMetrics: boolean = false;
     _isPreview: boolean;
 
+    /**
+     * The capture manager, used to manage captures (screenshots, videos, etc...).
+     */
+    _captureManager: CaptureManager;
```
Comment (4ian) [inline GDJS/Runtime/runtimegame.ts:194]: Should we allow this to be null?

```suggestion
    _captureManager: CaptureManager | null;
```

File: GDJS/Runtime/runtimegame.ts [hunk 15]
```diff
@@ -243,6 +251,10 @@ namespace gdjs {
       this._debuggerClient = gdjs.DebuggerClient
         ? new gdjs.DebuggerClient(this)
         : null;
+      this._captureManager = new gdjs.CaptureManager(
+        this._renderer,
+        this._options.captureOptions || {}
+      );
```
Comment (4ian) [inline GDJS/Runtime/runtimegame.ts:257]: (cf other comments, this can be done like the DebuggerClient:)

```suggestion
      this._captureManager = gdjs.CaptureManager ? new gdjs.CaptureManager(
        this._renderer,
        this._options.captureOptions || {}
      ) : null;
```

File: GDevelop.js/Bindings/Bindings.idl [hunk 16]
```diff
@@ -54,6 +54,11 @@ interface VectorObjectFolderOrObject {
     [Const] ObjectFolderOrObject at(unsigned long index);
 };
 
+interface VectorScreenshot {
+    unsigned long size();
+    [Value] Screenshot at(unsigned long index);
```
Comment (4ian) [inline GDevelop.js/Bindings/Bindings.idl:59]: Why not?

```suggestion
    [Const] Screenshot at(unsigned long index);
```

Returning a [Value] is risky because it uses a unique static variable. This means that doing:

const myScreenshot1 = myVector.at(0);
const myScreenshot2 = myVector.at(1);

//myScreenshot1 is now the SAME as myScreenshot2, because they both point to a unique, static object in C++ memory that is overriden every time it's called.

This also goes against what other VectorXXX are doing.

[Value] is almost a hack, if we could we would never use it.

File: GDevelop.js/Bindings/Bindings.idl [hunk 17]
```diff
@@ -1933,6 +1938,21 @@ interface QuickCustomizationVisibilitiesContainer {
     QuickCustomization_Visibility Get([Const] DOMString name);
 };
 
+interface Screenshot {
+    [Const] long GetTiming();
+    void SetTiming([Const] long timing);
```
Comment (4ian) [inline GDevelop.js/Bindings/Bindings.idl:1943]: No [Const] or [Ref] for primitive values.

```suggestion
    long GetTiming();
    void SetTiming(long timing);
```

File: GDevelop.js/Bindings/Bindings.idl [hunk 18]
```diff
@@ -3810,6 +3830,7 @@ interface PreviewExportOptions {
     [Ref] PreviewExportOptions SetGDevelopVersionWithHash([Const] DOMString gdevelopVersionWithHash);
     [Ref] PreviewExportOptions SetProjectTemplateSlug([Const] DOMString projectTemplateSlug);
     [Ref] PreviewExportOptions SetSourceGameId([Const] DOMString sourceGameId);
+    [Ref] PreviewExportOptions AddScreenshotCapture([Const] long timing, [Const] DOMString signedUrl, [Const] DOMString publicUrl);
```
Comment (4ian) [inline GDevelop.js/Bindings/Bindings.idl:3833]:
```suggestion
    [Ref] PreviewExportOptions AddScreenshotCapture(long delayTimeInMs, [Const] DOMString signedUrl, [Const] DOMString publicUrl);
```

File: Core/GDCore/IDE/CaptureOptions.h [hunk 19]
```diff
@@ -0,0 +1,48 @@
...
+  const long GetTiming() const { return timing; }
```
Comment (4ian) [inline Core/GDCore/IDE/CaptureOptions.h:15]:
```suggestion
 long GetTiming() const { return timing; }
```

(returning a primitive is not really needing any const, because you return a copy of it anyway, so the caller can do whatever they want with it, they won't modify the timing inside the class) (i.e: it's not a reference, so you get a copy)

File: GDevelop.js/Bindings/Bindings.idl (hunk 17 above)
Comment (4ian) [reply GDevelop.js/Bindings/Bindings.idl:1943]: (Const is mostly useless for these, they are passed by value, i.e: copied in memory. Const is useful when you pass a reference, and you don't want it to be modified. Often the case for a string or an object)

File: newIDE/app/src/ExportAndShare/BrowserExporters/BrowserS3PreviewLauncher/index.js [hunk 20]
```diff
@@ -219,10 +230,58 @@ export default class BrowserS3PreviewLauncher extends React.Component<
...
+                  screenshot => screenshot.publicUrl
+                );
+
+                // Check if they have been properly uploaded.
+                const responseUploadedScreenshotPublicUrls: Array<
+                  string | null
+                > = await Promise.all(
+                  screenshotPublicUrls.map(
+                    async (screenshotUrl): Promise<string | null> => {
+                      const response = await fetch(screenshotUrl, {
```
Comment (4ian) [inline newIDE/app/src/ExportAndShare/BrowserExporters/BrowserS3PreviewLauncher/index.js:258]: Can this throw? If so, it should be caught because we want to catch and consider this as a failure but not bubble up to the app.

File: newIDE/app/src/ExportAndShare/LocalExporters/LocalPreviewLauncher/index.js [hunk 21]
```diff
@@ -63,22 +65,68 @@ export default class LocalPreviewLauncher extends React.Component<
...
+          screenshot => screenshot.publicUrl
+        );
+
+        // Check if they have been properly uploaded.
+        const responseUploadedScreenshotPublicUrls: Array<
+          string | null
+        > = await Promise.all(
+          screenshotPublicUrls.map(
+            async (screenshotUrl): Promise<string | null> => {
+              const response = await fetch(screenshotUrl, { method: 'HEAD' });
```
Comment (4ian) [inline newIDE/app/src/ExportAndShare/LocalExporters/LocalPreviewLauncher/index.js:110]: Same as before, which makes me thing that this is enough to move this to a helper `const uploadedScreenshotPublicUrls = await uploadScreenshotsFromCaptureOptions(captureOptions);`

File: newIDE/app/src/ExportAndShare/LocalExporters/LocalPreviewLauncher/index.js [hunk 22]
```diff
@@ -63,22 +65,68 @@ export default class LocalPreviewLauncher extends React.Component<
...
-      alwaysOnTop: this.state.alwaysOnTop,
-      hideMenuBar: this.state.hideMenuBar,
-      numberOfWindows: this.state.numberOfWindows,
+      alwaysOnTop,
+      hideMenuBar,
+      numberOfWindows,
+      captureOptions,
+    });
+
+    ipcRenderer.removeAllListeners('preview-window-closed');
```
Comment (4ian) [inline newIDE/app/src/ExportAndShare/LocalExporters/LocalPreviewLauncher/index.js:95]: What if I launch two or 3 or more previews?

File: newIDE/app/src/MainFrame/EditorContainers/HomePage/LearnSection/TutorialsCategoryPage.js [hunk 23]
```diff
@@ -63,8 +63,11 @@ export const EducationCurriculum = ({
             );
           }
           items.push(
-            <div style={styles.sectionTitleContainer}>
-              <Text size="section-title" noMargin key={`section-${section}`}>
+            <div
+              style={styles.sectionTitleContainer}
+              key={`section-${section}`}
```
Comment (4ian) [inline newIDE/app/src/MainFrame/EditorContainers/HomePage/LearnSection/TutorialsCategoryPage.js:68]: Thanks!

File: newIDE/app/src/MainFrame/EditorContainers/BaseEditor.js [hunk 24]
```diff
@@ -94,6 +95,12 @@ export type RenderEditorContainerProps = {|
   askToCloseProject: () => Promise<boolean>,
   closeProject: () => Promise<void>,
 
+  // Games
+  games: ?Array<Game>,
+  fetchGames: () => Promise<void>,
+  onGameUpdated: (game: Game) => void,
+  gamesFetchingError: ?Error,
```
Comment (4ian) [inline newIDE/app/src/MainFrame/EditorContainers/BaseEditor.js:102]: Could this be passed as `gamesList: GamesList` type?
I'm afraid of these 4 properties becoming 5, then 6, then 7.

File: newIDE/app/src/MainFrame/index.js (hunk 3 above)
Comment (4ian) [reply newIDE/app/src/MainFrame/index.js:560]: See another comment, this could be a `gamesList: GamesList` type?

File: newIDE/app/src/MainFrame/index.js [hunk 25]
```diff
@@ -1630,6 +1655,43 @@ const MainFrame = (props: Props) => {
         currentProject
       );
 
+      const captureOptions = {};
```
Comment (4ian) [inline newIDE/app/src/MainFrame/index.js:1658]: Yet another 20 lines of logic in main frame: helper function?

File: Core/GDCore/IDE/CaptureOptions.cpp [hunk 26]
```diff
@@ -0,0 +1,18 @@
+/*
+ * GDevelop Core
+ * Copyright 2008-2016 Florian Rival (Florian.Rival@gmail.com). All rights
```
Comment (AlexandreSi) [inline Core/GDCore/IDE/CaptureOptions.cpp:3]:
```suggestion
 * Copyright 2008-2024 Florian Rival (Florian.Rival@gmail.com). All rights

```

File: newIDE/app/src/ExportAndShare/BrowserExporters/BrowserS3PreviewLauncher/index.js [hunk 27]
```diff
@@ -198,6 +198,17 @@ export default class BrowserS3PreviewLauncher extends React.Component<
           previewOptions.authenticatedPlayer.playerToken
         );
       }
+      if (previewOptions.captureOptions) {
+        if (previewOptions.captureOptions.screenshots) {
```
Comment (AlexandreSi) [inline newIDE/app/src/ExportAndShare/BrowserExporters/BrowserS3PreviewLauncher/index.js:202]: Merge ifs?

File: newIDE/app/src/ExportAndShare/LocalExporters/LocalPreviewLauncher/index.js (hunk 22 above)
Comment (ClementPasteau) [reply newIDE/app/src/ExportAndShare/LocalExporters/LocalPreviewLauncher/index.js:95]: This only works for the final preview launched indeed, I think that's fine?

File: Core/GDCore/IDE/CaptureOptions.h [hunk 28]
```diff
@@ -0,0 +1,50 @@
...
+
+class GD_CORE_API Screenshot {
+ public:
+  Screenshot();
+  virtual ~Screenshot() {};
+
+  void SetDelayTimeInSeconds(int delayTimeInMs_) {
+    delayTimeInMs = delayTimeInMs_;
+  }
+  const long GetDelayTimeInSeconds() const { return delayTimeInMs; }
```
Comment (4ian) [inline Core/GDCore/IDE/CaptureOptions.h:17]: - No need for const in the return type for primitives
- You return a long but the type is a int. Return an int.

File: newIDE/app/src/MainFrame/UseCapturesManager.js [hunk 29]
```diff
@@ -0,0 +1,149 @@
...
+          error
+        );
+      }
+
+      return captureOptions;
+    },
+    []
+  );
+
+  const onCaptureFinished = React.useCallback(
```
Comment (4ian) [inline newIDE/app/src/MainFrame/UseCapturesManager.js:70]: I like this better! All the logic to create signed url and then fetch them is here, rather then being splitted between here and the multiple files of the preview launchers.

**Review by 4ian — APPROVED** [review]
